src/Altinn.Notifications.Core/Services/OrderRequestService.cs

```diff
@@ -131,6 +118,46 @@ public async Task<NotificationOrderRequestResponse> RegisterNotificationOrder(No
         return lookupResult;
     }
 
+    private static List<string> GetMissingContactListIds(NotificationChannel channel, List<Recipient> recipients)
+    {
+        return channel switch
+        {
+            NotificationChannel.Email => recipients
+                               .Where(r => !r.AddressInfo.Exists(ap => ap.AddressType == AddressType.Email))
+                               .Select(r => r.OrganizationNumber ?? r.NationalIdentityNumber!)
+                               .ToList(),
+            NotificationChannel.Sms => recipients
+                                .Where(r => !r.AddressInfo.Exists(ap => ap.AddressType == AddressType.Sms))
+                                .Select(r => r.OrganizationNumber ?? r.NationalIdentityNumber!)
+                                .ToList(),
+            NotificationChannel.EmailPreferred or NotificationChannel.SmsPreferred => recipients
+                              .Where(r => !r.AddressInfo.Exists(ap => ap.AddressType == AddressType.Email || ap.AddressType == AddressType.Sms))
+                              .Select(r => r.OrganizationNumber ?? r.NationalIdentityNumber!)
+                              .ToList(),
+            _ => [],
+        };
+    }
+
+    private static List<Recipient> GetMissingContactRecipientList(NotificationChannel channel, List<Recipient> recipients)
+    {
+        return channel switch
+        {
+            NotificationChannel.Email => recipients
+                               .Where(r => !r.AddressInfo.Exists(ap => ap.AddressType == AddressType.Email))
+                               .Select(r => r.DeepCopy())
```

Comment:
Why is DeepCopy necessary? Will the original input ever be used assumed unchanged?

Reply:
If we don't deep copy the recipient info we find during the lookup would be persisted with the notification order. We do not want to persists this information until the send time. We first perform the lookup (triggered from line 50) and then we use the same object to generate the notification order to be persisted in the db (line 67)